File: sdks/go/examples/test_master_replica.go
package main

import (
	"fmt"
	"time"

	vectorizer "github.com/hivellm/vectorizer-sdk-go"
)

const (
	MASTER_URL  = "http://localhost:15002"
	REPLICA_URL = "http://localhost:17780"
	API_KEY     = "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiYWRtaW4iLCJ1c2VybmFtZSI6ImFkbWluIiwicm9sZXMiOlsiQWRtaW4iXSwiaWF0IjoxNzY0Mzk0MzI3LCJleHAiOjE3NjQzOTc5Mjd9.AnLPTgdHRfCFMdMp6VFemhcIZPUfpzjwB5r6xOkCxNQ"
)

func main() {
	fmt.Println("=== Go SDK Master/Replica Test ===\n")

	// 1. Test with hosts configuration
	fmt.Println("1. Creating client with hosts configuration...")
	client := vectorizer.NewClient(&vectorizer.Config{
		Hosts: &vectorizer.HostConfig{
			Master:   MASTER_URL,
			Replicas: []string{REPLICA_URL},
		},
		ReadPreference: vectorizer.ReadPreferenceReplica,
		APIKey:         API_KEY,
		Timeout:        30 * time.Second,
	})
	fmt.Println("   Client created with master/replica topology")

	// 2. Test health (read operation - should go to replica)
	fmt.Println("2. Testing Health() (read - should go to replica)...")
	err := client.Health()
	if err != nil {
		fmt.Printf("   Health failed: %v\n", err)
	} else {
		fmt.Println("   Health: OK")
	}

	// 3. Test listing collections (read operation)
	fmt.Println("3. Listing collections (read)...")
	collections, err := client.ListCollections()
	if err != nil {
		fmt.Printf("   List failed: %v\n", err)
	} else {
		fmt.Printf("   Found %d collections\n", len(collections))
	}

	// 4. Test WithMaster
	fmt.Println("4. Testing WithMaster()...")
	masterClient := client.WithMaster()
	err = masterClient.Health()
	if err != nil {
		fmt.Printf("   WithMaster health failed: %v\n", err)
	} else {
		fmt.Println("   Master Health: OK")
	}

	// 5. Test backward compatibility with single BaseURL
	fmt.Println("\n5. Testing backward compatibility (single BaseURL)...")
	singleClient := vectorizer.NewClient(&vectorizer.Config{
		BaseURL: MASTER_URL,
		APIKey:  API_KEY,
		Timeout: 30 * time.Second,
	})
	err = singleClient.Health()
	if err != nil {
		fmt.Printf("   Single URL failed: %v\n", err)
	} else {
		fmt.Println("   Single URL mode works: OK")
	}

	fmt.Println("\n=== Go SDK Test Complete ===")
}

File: sdks/rust/examples/test_master_replica.rs
//! Test Master/Replica routing functionality

use vectorizer_sdk::{ClientConfig, HostConfig, ReadPreference, VectorizerClient};

const MASTER_URL: &str = "http://localhost:15002";
const REPLICA_URL: &str = "http://localhost:17780";
const API_KEY: &str = "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiYWRtaW4iLCJ1c2VybmFtZSI6ImFkbWluIiwicm9sZXMiOlsiQWRtaW4iXSwiaWF0IjoxNzY0Mzk0MzI3LCJleHAiOjE3NjQzOTc5Mjd9.AnLPTgdHRfCFMdMp6VFemhcIZPUfpzjwB5r6xOkCxNQ";

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== Rust SDK Master/Replica Test ===\n");

    // 1. Test with hosts configuration
    println!("1. Creating client with hosts configuration...");
    let config = ClientConfig {
        hosts: Some(HostConfig {
            master: MASTER_URL.to_string(),
            replicas: vec![REPLICA_URL.to_string()],
        }),
        read_preference: Some(ReadPreference::Replica),
        api_key: Some(API_KEY.to_string()),
        ..Default::default()
    };

    let client = VectorizerClient::new(config)?;
    println!("   Client created with master/replica topology");

    // 2. Test health check (read operation - should go to replica)
    println!("2. Testing health check (read - should go to replica)...");
    match client.health_check().await {
        Ok(health) => println!("   Health status: {}", health.status),
        Err(e) => println!("   Health failed: {}", e),
    }

    // 3. Test listing collections (read operation)
    println!("3. Listing collections (read)...");
    match client.list_collections().await {
        Ok(collections) => println!("   Found {} collections", collections.len()),
        Err(e) => println!("   List failed: {}", e),
    }

    // 4. Test with_master callback
    println!("4. Testing with_master() callback...");
    match client
        .with_master(|master_client| async move {
            println!("   Inside with_master callback");
            let health = master_client.health_check().await?;
            Ok(health.status)
        })
        .await
    {
        Ok(status) => println!("   Master health: {}", status),
        Err(e) => println!("   with_master failed: {}", e),
    }

    // 5. Test backward compatibility with single base_url
    println!("\n5. Testing backward compatibility (single base_url)...");
    let single_config = ClientConfig {
        base_url: Some(MASTER_URL.to_string()),
        api_key: Some(API_KEY.to_string()),
        ..Default::default()
    };

    let single_client = VectorizerClient::new(single_config)?;
    match single_client.health_check().await {
        Ok(health) => println!("   Single URL mode works: {}", health.status),
        Err(e) => println!("   Single URL failed: {}", e),
    }

    println!("\n=== Rust SDK Test Complete ===");
    Ok(())
}
